PR for - Micrometer Integration - #64

laxmikantbpandhare opened the pull request:
Working on Timer and other remaining part. @metacosm and @csviri - Could you please verify the code for **Counter** and its implemented methods.

jmrodri commented:
@laxmikantbpandhare will this tie into the `deleteResource` method for decrementing the metric? https://github.com/java-operator-sdk/java-operator-sdk/blob/master/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/ResourceController.java#L28-L30

How will we ensure the increment occurs for the metrics we plan to support? Or is that for another follow on PR?

jmrodri commented:
I think that we should create a metrics  package in the `operator-framework-core/src/main/java/io/javaoperatorsdk/operator` or in `operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api` to house the metrics stuff. And I think CounterDetails should be in implementation detail of what we're trying to do. The issue lists a set of metrics we would like to expose by default. 

- number of controller executions
- number of events processed
- controller execution durations
- number of custom resource managed
- retry related metrics

So I think we should have a `metrics.ControllerExecution` which is an instance of a `CounterDetails`. We can then increment it for the user whenever a reconcile occurs. Similar to how controller-runtime is doing it here:  https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/internal/controller/metrics/metrics.go#L30-L33

```go
	ReconcileTotal = prometheus.NewCounterVec(prometheus.CounterOpts{
		Name: "controller_runtime_reconcile_total",
		Help: "Total number of reconciliations per controller",
	}, []string{"controller", "result"})
```

As you see in controller-runtime, there are a series of default structures that represent the default metrics they want to expose. I could see us having 5 classes representing the metrics we want.

Then in the `deleteResource` we would decrement the ControllerExecution.

adam-sandor commented on operator-framework-core/src/main/java/io/javaoperatorsdk/operator/utils/CounterDetails.java:
```diff
@@ -0,0 +1,40 @@
+
+public class CounterDetails {
+
+    int counter;
```
do these fields need to be package-private instead of private?

laxmikantbpandhare replied:
checking on this, we can make those field as a private.

metacosm replied:
I don't currently see what's the point of this class since it doesn't appear to be used but I would assume that it should be immutable: all fields should be private and final and there shouldn't be any setters (i.e. it shouldn't be possible to modify an instance, after it's been created)

laxmikantbpandhare replied:
The counter details class will be used when we call/use these utility methods. You can refer to sample **memcached-quarkus-operator** where I used this - https://github.com/laxmikantbpandhare/memcached-quarkus-operator/blob/main/src/main/java/com/example/MemcachedController.java

laxmikantbpandhare commented:
> @laxmikantbpandhare will this tie into the `deleteResource` method for decrementing the metric? https://github.com/java-operator-sdk/java-operator-sdk/blob/master/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/ResourceController.java#L28-L30
> 
> How will we ensure the increment occurs for the metrics we plan to support? Or is that for another follow on PR?

@jmrodri - I am still looking into the code of `java-operator-sdk` for **delete resource** functionality. I will add those changes as part of this PR only.

laxmikantbpandhare commented:
> @laxmikantbpandhare will this tie into the `deleteResource` method for decrementing the metric? https://github.com/java-operator-sdk/java-operator-sdk/blob/master/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/ResourceController.java#L28-L30
> 
> How will we ensure the increment occurs for the metrics we plan to support? Or is that for another follow on PR?

@jmrodri - We will not decrement any counter as **operator-sdk** does not decrement anything as such. After delete event, the reconciliation process happens thats why need to increment controller execution counter. The **operator-sdk** has two counters, **succeeded** and **failed**. If reconciliation process fails then **failed** counter increases. We will follow the same logic for our metrics part in **java-operator-sdk**.

jmrodri commented:
> @jmrodri - We will not decrement any counter as **operator-sdk** does not decrement anything as such. After delete event, the reconciliation process happens thats why need to increment controller execution counter. The **operator-sdk** has two counters, **succeeded** and **failed**. If reconciliation process fails then **failed** counter increases. We will follow the same logic for our metrics part in **java-operator-sdk**.

You are correct we only increment. Thanks.

jmrodri commented:
@laxmikantbpandhare also, this code should live in a `metrics` package.

laxmikantbpandhare commented:
> @laxmikantbpandhare also, this code should live in a `metrics` package.

@jmrodri - Sure, I will create `metrics` package and add those changes in it.

jmrodri commented:
Also remember that PRs should follow the conventional commit: https://github.com/java-operator-sdk/java-operator-sdk/blob/master/CONTRIBUTING.md#pull-request-process

laxmikantbpandhare commented:
> Also remember that PRs should follow the conventional commit: https://github.com/java-operator-sdk/java-operator-sdk/blob/master/CONTRIBUTING.md#pull-request-process

Modified commit as a conventional commit.

metacosm commented:
I've converted this PR to draft since it seems quite far from being finished. I cannot really review it in its current state because I need to see:
- which types of metrics will be provided
- how they will integrate with the SDK code
- which data will be recorded

laxmikantbpandhare commented:
@metacosm - I showed demo of metrics on our weekly meeting. Attached screenprint for your reference. 

<img width="1792" alt="Screen Shot 2021-08-17 at 12 00 21 AM" src="https://user-images.githubusercontent.com/47066536/129680524-390b0868-1e7f-4caf-9f0a-e8907eb4d8dc.png">

The current changes are just utility and I am planning to call those methods from reconciliation process.

metacosm commented:
On a design note: I'd like to see something similar to what is seen in https://github.com/ebullient/monster-combat/blob/main/core/src/main/java/dev/ebullient/dnd/combat/client/CombatMetrics.java. All metrics-related code is confined to a single class so that it doesn't "infect" the code and we can tweak what gets recorded and how in a single spot. The SDK code would call method of this class with application-specific results (e.g. `UpdateControl`) as opposed to things related to metrics, though we could add metrics-related information to these objects if needed. 
Also, we should only depend on `micrometer-core`, not `micrometer-registry-prometheus` so that we're not dependent on a given backend but users can choose whichever backend they want.

laxmikantbpandhare commented:
> On a design note: I'd like to see something similar to what is seen in https://github.com/ebullient/monster-combat/blob/main/core/src/main/java/dev/ebullient/dnd/combat/client/CombatMetrics.java. All metrics-related code is confined to a single class so that it doesn't "infect" the code and we can tweak what gets recorded and how in a single spot. The SDK code would call method of this class with application-specific results (e.g. `UpdateControl`) as opposed to things related to metrics, though we could add metrics-related information to these objects if needed.
> Also, we should only depend on `micrometer-core`, not `micrometer-registry-prometheus` so that we're not dependent on a given backend but users can choose whichever backend they want.

Yes, agree on this. I will update the code accordingly.

laxmikantbpandhare commented:
> On a design note: I'd like to see something similar to what is seen in https://github.com/ebullient/monster-combat/blob/main/core/src/main/java/dev/ebullient/dnd/combat/client/CombatMetrics.java. All metrics-related code is confined to a single class so that it doesn't "infect" the code and we can tweak what gets recorded and how in a single spot. The SDK code would call method of this class with application-specific results (e.g. `UpdateControl`) as opposed to things related to metrics, though we could add metrics-related information to these objects if needed.
> Also, we should only depend on `micrometer-core`, not `micrometer-registry-prometheus` so that we're not dependent on a given backend but users can choose whichever backend they want.

@metacosm - I completely removed CounterDetails class and added entire login in only one class. Please review the PR. About the second suggestion, `micrometer-core` contains only `SimpleMeterRegistry` and it is the simplest form of the registry. As per me, `MeterRegistry` is the best option. Below links explains more about it.

Links - 
https://micrometer.io/docs/installing
https://www.baeldung.com/micrometer#3-registry

metacosm commented:
We're not going to provide a registry implementation in the SDK: this will be up to users to decide which one to use. 
More precisely: the SDK will not be doing any actual metering, just provide the infrastructure for it to happen when a user selects a backend and activates it in their operator. `MeterRegistry` is the abstract class we should be coding against, not any implementation.

laxmikantbpandhare commented:
> We're not going to provide a registry implementation in the SDK: this will be up to users to decide which one to use.
> More precisely: the SDK will not be doing any actual metering, just provide the infrastructure for it to happen when a user selects a backend and activates it in their operator. `MeterRegistry` is the abstract class we should be coding against, not any implementation.

Ah! sorry it was my confusion. I updated dependency as said by you in `Quarkus Memcached Operator` and was getting `UnSatisfiedDepedency` for `micrometer-core`. It was my mistake. I updated the`pom.xml` file. Please review and let me know your comments.

adam-sandor commented:
Sorry @laxmikantbpandhare but I don't see where the actual functionality is supposed to be here... This is just an adding of a dependency and a simple counter class. Don't you want to actually use this in the code and count some things? Do you have a plan for what you want to measure?

laxmikantbpandhare commented:
> Sorry @laxmikantbpandhare but I don't see where the actual functionality is supposed to be here... This is just an adding of a dependency and a simple counter class. Don't you want to actually use this in the code and count some things? Do you have a plan for what you want to measure?

Exactly, as I said on the discord channel on operator-sdk-sync. Need some clarification on this.

metacosm commented:
Also note that your code is not properly formatted which is why CI fails.

metacosm commented:
@laxmikantbpandhare can you make your branch writable by maintainers, please? PRs should be opened in a way that maintainers can change stuff on it.

metacosm commented:
I've created #497 to show how I think we should go about the integration… Lots of work still to be done but should provide a basis for further discussions.

laxmikantbpandhare commented:
> @laxmikantbpandhare can you make your branch writable by maintainers, please? PRs should be opened in a way that maintainers can change stuff on it.

I already tick marked `Allow edits by maintainers`.

laxmikantbpandhare commented:
> I've created #497 to show how I think we should go about the integration… Lots of work still to be done but should provide a basis for further discussions.

@metacosm @adam-sandor @csviri - I created similar method for `deleteResource()`. Please review and let me know your comments.

metacosm commented on operator-framework-core/src/main/java/io/javaoperatorsdk/operator/Metrics.java:
```diff
@@ -0,0 +1,152 @@
+        registry.timer("operator.sdk.controllers.execution.delete", "controller", name);
+    try {
+      final var result = timer.record(() -> controller.deleteResource(resource, context));
+      String successType = "cr";
```
Do you really think that this success type makes sense for the delete operation?

laxmikantbpandhare replied:
```
      String successType = "notDelete";
      if (result == DeleteControl.DEFAULT_DELETE) {
        successType = "delete";
      }
 ```

I modified like this.

metacosm commented on operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/EventDispatcher.java:
```diff
@@ -153,7 +158,12 @@ private PostExecutionControl handleDelete(R resource, Context<R> context) {
         getName(resource),
         getVersion(resource));
     // todo: this is be executed in a try-catch statement, in case this fails
```
This comment is now obsolete.

laxmikantbpandhare replied:
Removed this comment.

metacosm commented on operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/EventDispatcher.java:
```diff
@@ -153,7 +158,12 @@ private PostExecutionControl handleDelete(R resource, Context<R> context) {
         getVersion(resource));
     // todo: this is be executed in a try-catch statement, in case this fails
-    DeleteControl deleteControl = controller.deleteResource(resource, context);
+    //    DeleteControl deleteControl = controller.deleteResource(resource, context);
```
Why keep this?

laxmikantbpandhare replied:
yes, removed this comment.

metacosm commented on operator-framework-core/src/test/java/io/javaoperatorsdk/operator/processing/EventDispatcherTest.java:
```diff
@@ -141,7 +147,7 @@ void callDeleteOnControllerIfMarkedForDeletionWhenNoFinalizerIsConfigured() {
         executionScopeWithCREvent(Watcher.Action.MODIFIED, testCustomResource));
 
-    verify(controller).deleteResource(eq(testCustomResource), any());
+    verify(controller, never()).deleteResource(eq(testCustomResource), any());
```
This isn't correct. Delete should be called.

laxmikantbpandhare replied:
Yes @metacosm , this is the test case where we are ignoring finalizer configuration. I modified the `configureToNotUseFinalizer()` method as below mentioned setup was missing when `configureToNotUseFinalizer()` creates a new `configuration`.

```
    when(configuration.getName()).thenReturn("EventDispatcherTestController");
    when(configService.getMetrics()).thenReturn(Metrics.NOOP);
    when(configuration.getConfigurationService()).thenReturn(configService);
```

I updated the PR, please let me know if anything is missing here.

metacosm commented:
On a strictly process-oriented basis: please do not request reviews for each commit. Finish the PR as you think it should be done and then we will be able to review it. Also, please leave the PR in draft mode until it's ready for review and merging, which isn't the case as this stage.

laxmikantbpandhare commented:
> On a strictly process-oriented basis: please do not request reviews for each commit. Finish the PR as you think it should be done and then we will be able to review it. Also, please leave the PR in draft mode until it's ready for review and merging, which isn't the case as this stage.

@metacosm - This process is completely new to us. I never requested review for each commit. Moving this in draft is something I never did. Is there any document where we have this entire process mentioned or it will be better if we elaborate during our weekly meeting.

csviri commented on operator-framework-core/src/main/java/io/javaoperatorsdk/operator/Metrics.java:
```diff
@@ -0,0 +1,152 @@
+import java.util.function.ToDoubleFunction;
+import java.util.function.ToLongFunction;
+
+public class Metrics {
```
For me this whole construct while wrapping the logic in a Metrics class is little bit smelly, not sure if there is a better way. But will take a look soon.

laxmikantbpandhare replied:
Initially, I wrote a separate class without wrapping it. @metacosm - Chris asked to have it like above wrapping the logic.

metacosm replied:
What's wrong with wrapping the logic, @csviri? More than wrapping, it's about measuring things that make sense from the operator's perspective as opposed to calling more metrics-oriented methods. Another aspect is that all metrics code is in a single spot which makes it easier to figure out what's measured and change it if needed.

csviri replied:
That the metrics code is in one place that is good. That we are mixing the application logic with metrics is not that nice. Those should be as loosely coupled as possible IMHO.

csviri replied:
I don't say I have a better answer for this now, just mentioning that design principle. So something to think about.

metacosm commented on operator-framework-core/src/main/java/io/javaoperatorsdk/operator/Metrics.java:
```diff
@@ -29,7 +29,10 @@ public Metrics(MeterRegistry registry) {
     final var name = configuration.getName();
     final var timer =
-        registry.timer("operator.sdk.controllers.execution.createOrUpdate", "controller", name);
+        Timer.builder("Controller Executions Timer")
```
Why was only this timer changed? What happened to the name and the tags? Also, please strive to create atomic, functional commits. A commit should stand by itself with a single functionality added / implemented as much as possible. As this commits stands, it's missing the delete counterpart which should be part of the same commit.

laxmikantbpandhare replied:
Yes, I will add same naming conventions and update it for Delete method as well.
